### ISSUE TITLE
Feature Request: Anti-slop / fine tuning of a model output in realtime / on the fly for output quality enhancement.

### ISSUE THREAD

**Issue by David-AU-github** [oobabooga/text-generation-webui#6443]
Feature Description

Basically this enhancement fixes model generation on the fly so to speak, and drastically improves the performance of any model
for specific tasks.

Although this is more involved than the "XTC" enhancement, this one is far stronger and will allow users to control the quality of generation of any model at the root level customized to their use case(s).

The occurs on the word, phrase level rather than per token level... roughly it forces the model to regenerate token(s) when generated token(s) match word(s)/phrase(s) in a json file. Like a live fire "regen" forcing the model to generate better/stronger output during generation.

Roughly is a live "proof" reading / editing during generation... but it seems to do more than that based on examples at the project website.

This may reduce T/S but increase quality output.
On smaller models this would allow them to perform well outside their "size" so to speak.

In some ways enhancement almost acts like "live fine tuning" of a model.

This is the project online for this enhancements:
https://github.com/sam-paech/antislop-sampler

Here is a detailed test at EQBENCH using this enhancement on a Gemma 9B top rated model:
[https://eqbench.com/results/creative-writing-v2/Gemma-2-Ataraxy-v2-9B%20[antislop].txt](https://eqbench.com/results/creative-writing-v2/Gemma-2-Ataraxy-v2-9B%20%5Bantislop%5D.txt)
Motivation

Drastic improvement of output quality on the fly in real time, tunable by the use / use case.
Smaller and mid size model improvement across the board.
Larger models would also benefit - could result in drastic leap in coherence / generation and maybe logic solving improvement.

This could match or exceed closed source model performance depending on implementation.
The quality enhancement at small / mid-sized models can not be overstated.

Roughly allows fine tuning of a model at the user / case level.

On the user level: Far fewer (if any) "regens" to get good output quality.
That all by itself is a game changer.
This would drastically improve user / end user experience across the board.
Possible Implementation

Based on the project online and noted ; one or more "user defined" text files / json for specific use case(s).
This could be a "config.json" at the "source level" before quanting and/or post use / post quant.

Option to download hugging face, one or more of these files to augment usage of model(s) much like a dataset is used to fine tune a model.

I would suggest implementing this as a minimum to allow one or more files to be selected at the quantize (similar to embedding "config.json") in the GGUF step and/or inference step(s).

IE:
--optimize creative.json

OR when you "run" the model:

--enhance creative.json

Which would then run "creative.json" live as per project noted here during output generation:
https://github.com/sam-paech/antislop-sampler

NOTE: Also submitted this at llamacpp , no idea if it will be accepted.

**Comment by FartyPants** [comment]
I think this might work better as an extension. There was an example to change logits I remember, that would be easy to adopt to this as it simply uses json table to lower probability of slop words.

**Comment by Feldherren** [comment]
Support for antislop sampler functionality like this would go a long way to solving issues I have with the latest generation of models, whether it's an extension or not. Some other options for locally loading models already support it.

**Comment by Tomobobo710** [comment]
Man I just started to play with this in the Koboldcpp phrase banning section and it's amazing bro, much better than the token banning feature. Are you saying you're writing an extension for text-gen-ui?

**Comment by David-AU-github** [comment]
Currently it is a SillyTavern one, experimental.

Here is an early BETA (I am "DavidAU") - this is a real time streaming auto-modification:

https://huggingface.co/DavidAU/AI_Autocorrect__Auto-Creative-Enhancement__Auto-Low-Quant-Optimization__gguf-exl2-hqq-SOFTWARE

Anti-Slop is not in this version yet - although versions exist in the "lab", and the BETA has all the framework (and tested)" - and I have tested several different types of "Anti-slop" as well as throwing the model a curve ball(s) too. Kind of like "Auto-correct" but much more directed.

I have created versions with Auto-Correct, Creative Enhancement, Curveballs, triggers, and multi-anti-slops (IE word, phrase, specific types, trigger etc) all running concurrently. Getting the priorities correct was a major hurdle.

There is an issue with Anti-slop(s) however ... that the content ie words/phrases you need for good anti-slop **vary on genre / use case.** Making it use case specific will increase its effect, while negating "false" anti-slop actions.

I have a solution for this, just working on the coding which will happen next week.

One of the core goals -> Get the anti-slop(s) to force the model to make better decisions => Better generation .
The system ENFORCES all of these directives, but unlike anti-slop/token banning this happens at multiple levels.

Side note: These extensions are at the "API" end so to speak, and can easily be converted into other languages / installed into other systems. They are written in standard JS with no dependencies.

Translation to Python should be very straightforward.

**Comment by Feldherren** [comment]
> that the content ie words/phrases you need for good anti-slop vary on genre / use case. 
Isn't that why most implementations of it so far take a user-supplied set of phrases to watch for? I think the linked one even takes regex. Everyone's going to dislike different phrases, and the undesirable phrases are going to differ heavily based on the LLM model used.

**Comment by David-AU-github** [comment]
Agreed ; user options and...

-> Default set (on/off)

-> Also a more advance option that sets itself up automatically. 
IE: Self creates the anti slop "sets" so to speak based on use case then applies these sets automatically in addition to user and/or default sets.

The other issue-> Context. ; ie: in Writing, "slop" is ok in "thoughts", "dialog" but not in narration/desc/beats etc etc.
